module/zfs/mmp.c

```diff
@@ -258,27 +235,40 @@ mmp_random_leaf_impl(vdev_t *vd, int *fail_mask)
  * MMP_FAIL_NOT_WRITABLE
  * If set, one or more vdevs are not writeable.  The children of those vdevs
  * were not examined.
- *
- * Assuming in_vd points to a tree, a random subtree will be chosen to start.
- * That subtree, and successive ones, will be walked until a usable leaf has
- * been found, or all subtrees have been examined (except that the children of
- * un-writeable vdevs are not examined).
- *
- * If the leaf vdevs in the tree are healthy, the distribution of returned leaf
- * vdevs will be even.  If there are unhealthy leaves, the following leaves
- * (child_index % index_children) will be chosen more often.
  */
 
 static int
-mmp_random_leaf(vdev_t *in_vd, vdev_t **out_vd)
+mmp_next_leaf(spa_t *spa, vdev_t **current_leaf)
 {
-	int error_mask = 0;
-	vdev_t *vd = mmp_random_leaf_impl(in_vd, &error_mask);
+	vdev_t *leaf;
+	boolean_t found = B_FALSE;
+	int fail_mask = 0;
+
+	ASSERT(current_leaf != NULL);
+	ASSERT(list_link_active(&spa->spa_leaf_list.list_head) == B_TRUE);
```

Comment:
I'm not sure i understand this second assertion: `ASSERT(list_link_active(&spa->spa_leaf_list.list_head) == B_TRUE)`.

Reply:
It's asserting that spa_leaf_list was initialized with list_create().   It may be unnecessary, I suspect things would blow up all over if it were not.  But I do not see that there is an equivalent assertion in the basic user space list functions, like list_insert_head() et al., which was my reason for putting it there.